### ISSUE TITLE
Combined Squeeze-Excite + [Ghost] BatchNorm

### ISSUE THREAD

**Issue by dblalock** [mosaicml/composer#745]
## New Method

This method would replace each batchnorm module with a new "SE-BatchNorm" module. This module would have input-dependent channel scaling like [squeeze-excite](https://docs.mosaicml.com/en/latest/method_cards/squeeze_excite.html), and also support using a ghost batch size like [GhostBatchNorm](https://docs.mosaicml.com/en/latest/method_cards/ghost_batchnorm.html). 

Essentially, the bottleneck in both batchnorm and squeeze-excite is computing channel statistics. If you call these methods separately, you're doing redundant work. With a shared call to `torch.var_mean(input, dim=(-2, -1))`, the combination of the two methods would be much cheaper during training. Also, since we're manually computing the statistics instead of calling `F.batch_norm`, it's easy to add ghost batching while we're at it.

## Motivation

BatchNorm is a staple of CNN training. And we've seen Squeeze-Excite consistently improve time-to-train by a large margin, despite slowing down the model. If we could add in squeeze excite (and ghost batchnorm) with less speed cost, time-to-train would improve even more. Although note that this fusion wouldn't make SE faster at inference time.

This would also provide a cleaner story for where to add SE within the model than we currently have. At present, we add it after every Conv2d with sufficient channel count, which was a necessary deviation from the paper in order to make the surgery work for any CNN.

## [Optional] Implementation

```
def forward(self, input):
    sample_channel_vars, sample_channel_means = torch.varmean(input, dim=(-2, -1))

    # compute batchnorm batch statistics
    batch_channel_vars = sample_channel_vars.reshape(self.ghost_batch_size, ...).mean(dim=0)
    batch_channel_means = sample_channel_means.reshape(self.ghost_batch_size, ...).mean(dim=0)

    # compute SE channel attention
    se_channel_weights = self.se_mlp(sample_channel_means)

    # combine [ghost] batchnorm affine transform with SE linear transform
    output = input.reshape(self.ghost_batch_size, ...) - batch_channel_means
    output = output.reshape(input.shape)
    return output * (se_channel_weights / batch_channel_vars)  # parentheses critical
```
The above won't quite work, but hopefully conveys the idea. The main points are that:

1. A single `torch.var_mean` call is shared for everything
2. The SE channel weighting and the batchnorm channel variances are combined together *before* being broadcast across all the spatial positions.

Ideally, the implementation would also ensure that a regular BatchNorm call happens when the model is in eval mode, so that TensorRT, etc, will be able to fuse the batchnorm into adjacent ops. Although this probably doesn't matter much when we're computing the SE statistics regardless.

**Comment by b-thebest** [comment]
Can I work on this?

**Comment by dblalock** [comment]
@b-thebest Sure, go for it!

**Comment by mvpatel2000** [comment]
Closing. Tracking elsewhere as low pri. We're open to community suggestions!
